### PR TITLE
NIFI-13649 Ensure cluster node API address is reachable before markin…

### PR DESCRIPTION
**Pull request by bbende**
…g CONNECTED

# Summary

[NIFI-13649](https://issues.apache.org/jira/browse/NIFI-13649)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java [hunk 1]
```diff
@@ -296,6 +305,16 @@ private void processHeartbeat(final NodeHeartbeat heartbeat) {
         clusterCoordinator.validateHeartbeat(heartbeat);
     }
 
+    private boolean isNodeApiReachable(final NodeIdentifier nodeIdentifier) {
+        try {
+            final InetAddress nodeApiAddress = InetAddress.getByName(nodeIdentifier.getApiAddress());
+            return nodeApiAddress.isReachable((int) NODE_API_TIMEOUT.toMillis());
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java:311]: The [isReachable](https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html#isReachable-int-) method can behave differently depending on the platform, sometimes using ICMP, and sometimes using TCP port 7. For a more reliable check, seems worth considering making a socket connection to the node port, which would ensure basic TCP connectivity.

File: nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java [hunk 2]
```diff
@@ -288,6 +292,11 @@ private void processHeartbeat(final NodeHeartbeat heartbeat) {
                 return;
             }
 
+            if (!isNodeApiReachable(nodeId)) {
+                logger.info("Received a connection request from {}, but the node's API address is not reachable, will not connect node to the cluster yet", nodeId);
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java:296]: What do you think about adjusting the wording?
```suggestion
                logger.info("Node API Address [{}] not reachable: cluster connection request deferred pending successful network connection", nodeId);
```

**Review by exceptionfactory — COMMENTED** [review]
Thanks for introducing the improvements here @bbende. I noted a concern regarding the `isReachable()` method behavior, and I think an alternative approach using a basic Socket connection would be more reliable across platforms.

**Comment by bbende** [comment]
Thanks @exceptionfactory , pushed an update to address the feedback

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @bbende, the latest version looks good! +1 merging

**Pull request merged**